In [2]:
import os
# Print current working directory
print("Current working directory:", os.getcwd())

Current working directory: c:\Users\skati\MSc Bioinformatics BBK\Antibody-Constant-Region-Modelling-Pipeline\Scripts


In [16]:
import os

# Set the project root using absolute path and adding ".." to 
# move up 1 level from .../Modelling-Pipeline/Scripts -> .../Modelling-Pipeline
project_root = os.path.abspath(os.path.join(os.getcwd(), ".."))

# List of required directories
required_dirs = [
    "VCAb_data",
    "fasta_sequences",
    "alignments",
    "pir_files",
    "atom_files",
    "models",
    "pickles"
]

# Create each directory if it doesn't exist
for dir_name in required_dirs:
    os.makedirs(os.path.join(project_root, dir_name), exist_ok=True)

In [1]:
import pandas as pd
import my_run_info
#from my_run_info import models_out_dir


# Pandas to read csv file as a dataframe
df = pd.read_csv(my_run_info.VCAb_dir)

# Simplify the naming conventions in certain columns using str.extract(r"regex_expression")
# This captures the matched word at the start of string, stopping at the first non-word character "("
df["H_isotype_clean"] = df["Htype"].str.extract(r"^(\w+)")
df["L_isotype_clean"] = df["Ltype"].str.extract(r"^(\w+)")

# Reduce dataframe size down to just the key columns (especially the "_clean" ones)
columns_to_keep = [
    "pdb","Hchain", "Lchain", "H_seq", "L_seq", "H_coordinate_seq", "L_coordinate_seq", 
    "H_PDB_numbering", "L_PDB_numbering", "pdb_H_VC_Boundary", "pdb_L_VC_Boundary", "title", "release_date","method", "resolution", 
    "carbohydrate", "HC_species", "H_isotype_clean", "L_isotype_clean", "HC_coordinate_seq", 
    "LC_coordinate_seq", "HV_seq", "LV_seq","disulfide_bond"
]
df_keep = df[columns_to_keep]

# Filter down to rows of interest with conditions. Using wrapped conditions with & 
# `[(cond_1) & (cond_2) & ...etc]`.
df_only_human_and_kappa_chain = df_keep[
    (df_keep["HC_species"] == "homo_sapiens") & 
    (df_keep["L_isotype_clean"].str.contains("kappa", case=False))
]

# Visualise the data
print(f"Rows, columns: {df_only_human_and_kappa_chain.shape}")
df_only_human_and_kappa_chain.head(5)


Rows, columns: (3767, 24)


pdb Hchain Lchain                                              H_seq  \
5   1a4j    B;H    A;L  QVQLLESGPELKKPGETVKISCKASGYTFTNYGMNWVKQAPGKGLK...   
6   1a4k      B      A  QVQLLESGPELKKPGETVKISCKASGYTFTNYGMNWVKQAPGKGLK...   
7   1a4k      H      L  QVQLLESGPELKKPGETVKISCKASGYTFTNYGMNWVKQAPGKGLK...   
11  1ad0    B;D    A;C  EVQLLESGGGLVQPGGSLRLSCATSGFTFTDYYMNWVRQAPGKGLE...   
12  1ad9    B;H    A;L  EIQLVQSGAEVKKPGSSVKVSCKASGYTFTDYYINWMRQAPGQGLE...   

                                                L_seq  \
5   ELVMTQTPLSLPVSLGDQASISCRSSQSLVHSNGNTYLHWYLQKPG...   
6   ELVMTQTPLSLPVSLGDQASISCRSSQSLLHSNGNTYLHWYLQKPG...   
7   ELVMTQTPLSLPVSLGDQASISCRSSQSLLHSNGNTYLHWYLQKPG...   
11  QTVLTQSPSSLSVSVGDRVTITCRASSSVTYIHWYQQKPGLAPKSL...   
12  DIQMTQSPSTLSASVGDRVTITCRSSKSLLHSNGDTFLYWFQQKPG...   

                                     H_coordinate_seq  \
5   QVQLLESGPELKKPGETVKISCKASGYTFTNYGMNWVKQAPGKGLK...   
6   QVQLLESGPELKKPGETVKISCKASGYTFTNYGMNWVKQAPGKGLK...   
7   QVQLLESGPELKKPGETVKISCKASGYTFTNYGMNWVKQAPGKGLK...   
11  EVQLLESGGGLVQPGGSLRLSCATSGFTFTDYYMNWVRQAPGKGLE...   
12  EIQLVQSGAEVKKPGSSVKVSCKASGYTFTDYYINWMRQAPGQGLE...   

                                     L_coordinate_seq  \
5   ELVMTQTPLSLPVSLGDQASISCRSSQSLVHSNGNTYLHWYLQKPG...   
6   ELVMTQTPLSLPVSLGDQASISCRSSQSLLHSNGNTYLHWYLQKPG...   
7   ELVMTQTPLSLPVSLGDQASISCRSSQSLLHSNGNTYLHWYLQKPG...   
11  QTVLTQSPSSLSVSVGDRVTITCRASSSVTYIHWYQQKPGLAPKSL...   
12  DIQMTQSPSTLSASVGDRVTITCRSSKSLLHSNGDTFLYWFQQKPG...   

                                      H_PDB_numbering  \
5   1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...   
6   1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...   
7   1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...   
11  1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...   
12  1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...   

                                      L_PDB_numbering pdb_H_VC_Boundary  ...  \
5   1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...               120  ...   
6   1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...               114  ...   
7   1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...               114  ...   
11  1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...               114  ...   
12  1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...               114  ...   

   resolution carbohydrate    HC_species H_isotype_clean  L_isotype_clean  \
5         2.1          NaN  homo_sapiens            IgG1            kappa   
6         2.4          NaN  homo_sapiens            IgG1            kappa   
7         2.4          NaN  homo_sapiens            IgG1            kappa   
11        2.5          NaN  homo_sapiens            IgG4            kappa   
12        2.8          NaN  homo_sapiens            IgG4            kappa   

                                    HC_coordinate_seq  \
5   ASTKGPSVFPLAPSSKSTSGGTAALGCLVKDYFPEPVTVSWNSGAL...   
6   ASTKGPSVFPLAPSSKSTSGGTAALGCLVKDYFPEPVTVSWNSGAL...   
7   ASTKGPSVFPLAPSSKSTSGGTAALGCLVKDYFPEPVTVSWNSGAL...   
11  ASTKGPSVFPLAPCSRSTSESTAALGCLVKDYFPEPVTVSWNSGAL...   
12  ASTKGPSVFPLAPCSRSTSESTAALGCLVKDYFPEPVTVSWNSGAL...   

                                    LC_coordinate_seq  \
5   RTVAAPSVFIFPPSDEQLKSGTASVVCLLNNFYPREAKVQWKVDNA...   
6   RTVAAPSVFIFPPSDEQLKSGTASVVCLLNNFYPREAKVQWKVDNA...   
7   RTVAAPSVFIFPPSDEQLKSGTASVVCLLNNFYPREAKVQWKVDNA...   
11  RTVAAPSVFIFPPSDEQLKSGTASVVCLLNNFYPREAKVQWKVDNA...   
12  RTVAAPSVFIFPPSDEQLKSGTASVVCLLNNFYPREAKVQWKVDNA...   

                                               HV_seq  \
5   QVQLLESGPELKKPGETVKISCKASGYTFTNYGMNWVKQAPGKGLK...   
6   QVQLLESGPELKKPGETVKISCKASGYTFTNYGMNWVKQAPGKGLK...   
7   QVQLLESGPELKKPGETVKISCKASGYTFTNYGMNWVKQAPGKGLK...   
11  EVQLLESGGGLVQPGGSLRLSCATSGFTFTDYYMNWVRQAPGKGLE...   
12  EIQLVQSGAEVKKPGSSVKVSCKASGYTFTDYYINWMRQAPGQGLE...   

                                               LV_seq  \
5   ELVMTQTPLSLPVSLGDQASISCRSSQSLVHSNGNTYLHWYLQKPG...   
6   ELVMTQTPLSLPVSLGDQASISCRSSQSLLHSNGNTYLHWYLQKPG...   
7   ELVMTQTPLSLPVSLGDQASIS

In [2]:
# Create a list of rows of interest (can be by PDB ID or Heavy Chain Type)

# Define the templates of interest
# - Variable Heavy & Light template
v_template = "1n8z"
# - Constant Heavy & Light template
c_template = "3ncj"

# Filter to template rows. Filters only rows where matching pdb is True. 
# NOTE: df_filtered is now and independant copy of df_only_human_and_kappa
df_filtered = df_only_human_and_kappa_chain[
    df_only_human_and_kappa_chain["pdb"].isin([v_template, c_template])
].copy()

# Add a 'template' annotation column
df_filtered.insert(loc=1, column="template", value=None)
df_filtered.loc[df_filtered["pdb"] == v_template, "template"] = "v_template"
df_filtered.loc[df_filtered["pdb"] == c_template, "template"] = "c_template"

df_filtered

pdb    template Hchain Lchain  \
242   1n8z  v_template      B      A   
1062  3ncj  c_template      H      L   

                                                  H_seq  \
242   EVQLVESGGGLVQPGGSLRLSCAASGFNIKDTYIHWVRQAPGKGLE...   
1062  EVQLVQSGAEVKKPGESLKISCKGSGYSFTNYWISWVRQMPGKGLE...   

                                                  L_seq  \
242   DIQMTQSPSSLSASVGDRVTITCRASQDVNTAVAWYQQKPGKAPKL...   
1062  DIQMTQSPSSLSASVGDRVTITCRASQSIGLYLAWYQQKPGKAPKL...   

                                       H_coordinate_seq  \
242   EVQLVESGGGLVQPGGSLRLSCAASGFNIKDTYIHWVRQAPGKGLE...   
1062  EVQLVQSGAEVKKPGESLKISCKGSGYSFTNYWISWVRQMPGKGLE...   

                                       L_coordinate_seq  \
242   DIQMTQSPSSLSASVGDRVTITCRASQDVNTAVAWYQQKPGKAPKL...   
1062  DIQMTQSPSSLSASVGDRVTITCRASQSIGLYLAWYQQKPGKAPKL...   

                                        H_PDB_numbering  \
242   1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...   
1062  1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...   

                                        L_PDB_numbering  ... resolution  \
242   1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...  ...       2.52   
1062  1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...  ...       1.60   

     carbohydrate    HC_species  H_isotype_clean L_isotype_clean  \
242           NaN  homo_sapiens             IgG1           kappa   
1062          NaN  homo_sapiens             IgG4           kappa   

                                      HC_coordinate_seq  \
242   ASTKGPSVFPLAPSSKSTSGGTAALGCLVKDYFPEPVTVSWNSGAL...   
1062  ASTKGPSVFPLAPCSRSTSESTAALGCLVKDYFPEPVTVSWNSGAL...   

                                      LC_coordinate_seq  \
242   RTVAAPSVFIFPPSDEQLKSGTASVVCLLNNFYPREAKVQWKVDNA...   
1062  RTVAAPSVFIFPPSDEQLKSGTASVVCLLNNFYPREAKVQWKVDNA...   

                                                 HV_seq  \
242   EVQLVESGGGLVQPGGSLRLSCAASGFNIKDTYIHWVRQAPGKGLE...   
1062  EVQLVQSGAEVKKPGESLKISCKGSGYSFTNYWISWVRQMPGKGLE...   

                                                 LV_seq  \
242   DIQMTQSPSSLSASVGDRVTITCRASQDVNTAVAWYQQKPGKAPKL...   
1062  DIQMTQSPSSLSASVGDRVTITCRASQSIGLYLAWYQQKPGKAPKL...   

                                         disulfide_bond  
242   CYS147(B)-CYS203(B):6.59, CYS134(A)-CYS194(A):...  
1062  CYS127(H)-CYS214(L):4.72, CYS140(H)-CYS196(H):...  

[2 rows x 25 columns]

In [ ]:
# DO NOT USE

    # Is "LVTVSS" present in "H_coordinate_seq"? If not True "H_coordinate_seq has atypical VH-boundary"
    # Is "LVTVSS" present in "HV_coordinate_seq"? If not true "Not present in HV_coordinate_seq
    # Is "LVTVSS" the last 6 characters in "HV_coordinate_seq"?
    # How many residues appear after "LVTVSS"? --> Store these as overlap = str(), along with len(overlap)
    # If overlap residues == the first len(overlap) residues of HC_coordinate_seq, then remove them from

def template_parse_check(df):
    """Checks that atom coordinates of the .cif template file exactly match those in the VCAb listing."""
    
    # Using iterrows for stable behaviour in pandas
    for idx, row in df.iterrows():

        # Define sequence lengths                    
        heavy_full_length = len(str(row["H_seq"]))
        heavy_cif_length = len(str(row["H_coordinate_seq"]))
            
        light_full_length = len(str(row["L_seq"]))
        light_cif_length = len(str(row["L_coordinate_seq"]))

        # Quick Length Check
        if heavy_full_length == heavy_cif_length:
            print(f"{row["pdb"]} heavy chain: full sequence matches template cif ({heavy_cif_length})")
        else:
            print(f"{row["pdb"]} heavy chain: full sequence ({heavy_full_length}) differs from template cif ({heavy_cif_length})")
            
        if light_full_length == light_cif_length:
            print(f"{row["pdb"]} light chain: full sequence matches template cif ({light_cif_length})")
        else:
            print(f"{row["pdb"]} light chain: full sequence ({light_full_length}) differs from template cif ({light_cif_length})")
            

    
template_parse_check(df_filtered)

1n8z heavy chain: full sequence matches template cif (220)
1n8z light chain: full sequence matches template cif (214)
2agj heavy chain: full sequence (226) differs from template cif (225)
2agj light chain: full sequence matches template cif (215)


In [ ]:
# Can be called separately for each template
def zip_template_cif(df: pd.DataFrame, pdb_code: str):
    """Takes in a filtered DataFrame and PDB code. Zips residues with 
    PDB coordinates to their respective PDB numbering, 
    only residues appearing in the x-ray structure are included. 
    Outputs a list of tuples for residues and their numbering."""
    
    # Using iterrows for stable behaviour in pandas
    for idx, row in df.iterrows():

        if row["pdb"] == pdb_code:
            
            # Parse .cif Heavy Chain 
            heavy_cif_residue_list = list(row["H_coordinate_seq"])
            # Remove duplicate numbering with letter annotations eg: "100A"
            heavy_cif_numbering_list = [int(res) for res in row["H_PDB_numbering"].split(",") if res.isdigit()]
            heavy_cif_VC_boundary = int(row["pdb_H_VC_Boundary"])
        
            # Parse .cif Light Chain
            light_cif_residue_list = list(row["L_coordinate_seq"])
            # Remove duplicate numbering with letter annotations eg: "100A"
            light_cif_numbering_list = [int(res) for res in row["L_PDB_numbering"].split(",") if res.isdigit()]
            light_cif_VC_boundary = int(row["pdb_L_VC_Boundary"])

            # Create heavy/light chain zips
            heavy_zip = (list(zip(heavy_cif_residue_list, heavy_cif_numbering_list)))
            light_zip = (list(zip(light_cif_residue_list, light_cif_numbering_list)))
            
            print(f"Zipped light chain length: {len(light_zip)}")
            print(f"Zipped heavy chain length: {len(heavy_zip)}")
            print(f"Light chain VC boundary {light_cif_VC_boundary}th residue")
            print(f"Heavy chain VC boundary {heavy_cif_VC_boundary}th residue")

    return (light_zip, heavy_zip, light_cif_VC_boundary, heavy_cif_VC_boundary)

zip_template_cif(df_filtered, c_template)


Zipped light chain length: 214
Zipped heavy chain length: 217
Light chain VC boundary 108th residue
Heavy chain VC boundary 114th residue


([('D', 1),
  ('I', 2),
  ('Q', 3),
  ('M', 4),
  ('T', 5),
  ('Q', 6),
  ('S', 7),
  ('P', 8),
  ('S', 9),
  ('S', 10),
  ('L', 11),
  ('S', 12),
  ('A', 13),
  ('S', 14),
  ('V', 15),
  ('G', 16),
  ('D', 17),
  ('R', 18),
  ('V', 19),
  ('T', 20),
  ('I', 21),
  ('T', 22),
  ('C', 23),
  ('R', 24),
  ('A', 25),
  ('S', 26),
  ('Q', 27),
  ('S', 28),
  ('I', 29),
  ('G', 30),
  ('L', 31),
  ('Y', 32),
  ('L', 33),
  ('A', 34),
  ('W', 35),
  ('Y', 36),
  ('Q', 37),
  ('Q', 38),
  ('K', 39),
  ('P', 40),
  ('G', 41),
  ('K', 42),
  ('A', 43),
  ('P', 44),
  ('K', 45),
  ('L', 46),
  ('L', 47),
  ('I', 48),
  ('Y', 49),
  ('A', 50),
  ('A', 51),
  ('S', 52),
  ('S', 53),
  ('L', 54),
  ('Q', 55),
  ('S', 56),
  ('G', 57),
  ('V', 58),
  ('P', 59),
  ('S', 60),
  ('R', 61),
  ('F', 62),
  ('S', 63),
  ('G', 64),
  ('S', 65),
  ('G', 66),
  ('S', 67),
  ('G', 68),
  ('T', 69),
  ('D', 70),
  ('F', 71),
  ('T', 72),
  ('L', 73),
  ('T', 74),
  ('I', 75),
  ('S', 76),
  ('S', 77),
  ('L', 

In [11]:
def get_constant_region(light_zip, heavy_zip, light_boundary, heavy_boundary):
    """Extracts the tuple after the Variable-Constant boundary within a heavy/light chain zip."""

    # Expression format: `(return this for (item1, item2) in iterable if condition)`
    # expression - what we want to return
    # for ... in ... - the loop part
    # if ... - optional condition

    # Find index of the boundary tuple in each chain
    light_boundary_index = next(i for i, (_, num) in enumerate(light_zip) if num == light_boundary)
    heavy_boundary_index = next(i for i, (_, num) in enumerate(heavy_zip) if num == heavy_boundary)

    #print(f"Light Boundary tuple index: {light_boundary_index}")
    #print(f"Heavy Boundary tuple index: {heavy_boundary_index}")

    # Get all tuples after the boundary
    light_post_boundary = light_zip[light_boundary_index + 1:]
    heavy_post_boundary = heavy_zip[heavy_boundary_index + 1:]  

    return (light_post_boundary, heavy_post_boundary)

# Use * to unpack the tuple from zip function
cl_zip, ch_zip = get_constant_region(*zip_template_cif(df_filtered, c_template))

Zipped light chain length: 214
Zipped heavy chain length: 217
Light chain VC boundary 108th residue
Heavy chain VC boundary 114th residue


In [12]:
def get_variable_region(light_zip, heavy_zip, light_boundary, heavy_boundary):
    """Extracts the tuple before the Variable-Constant boundary within a heavy/light chain zip."""

    # Find index of the boundary tuple in each chain
    light_boundary_index = next(i for i, (_, num) in enumerate(light_zip) if num == light_boundary)
    heavy_boundary_index = next(i for i, (_, num) in enumerate(heavy_zip) if num == heavy_boundary)

    #print(f"Light Boundary tuple index: {light_boundary_index}")
    #print(f"Heavy Boundary tuple index: {heavy_boundary_index}")

    # Get all tuples BEFORE AND INCLUDING the boundary
    light_pre_boundary = light_zip[:light_boundary_index + 1]
    heavy_pre_boundary = heavy_zip[:heavy_boundary_index + 1]

    return (light_pre_boundary, heavy_pre_boundary)

# Use * to unpack the tuple from zip function
vl_zip, vh_zip = get_variable_region(*zip_template_cif(df_filtered, v_template))


Zipped light chain length: 214
Zipped heavy chain length: 220
Light chain VC boundary 108th residue
Heavy chain VC boundary 121th residue


In [13]:
# The recombinant antibody must only combine V sequence of v_template + C sequence of c_template

def make_recombinant_seqs(vl_zip, vh_zip, cl_zip, ch_zip):
    """Combines the variable and constant regions of 2 different template sequences.
    Takes 'zipped' (residue-position sequence) chains, joins VL to CL and VH to CH.
    Outputs a tuple of two strings (light_sequence, heavy_sequence)."""

    # Make variable region sequence
    vl_str = ''.join(res for res, _ in vl_zip)
    vh_str = ''.join(res for res, _ in vh_zip)

    # Make constant region sequence
    cl_str = ''.join(res for res, _ in cl_zip)
    ch_str = ''.join(res for res, _ in ch_zip)

    recombinant_seq_light = vl_str + cl_str
    recombinant_seq_heavy = vh_str + ch_str

    print(f"Recombinant Sequence (light): {recombinant_seq_light}")
    print(f"Recombinant Sequence (heavy): {recombinant_seq_heavy}")

    return (recombinant_seq_light, recombinant_seq_heavy)

recombinant_seq_light, recombinant_seq_heavy = make_recombinant_seqs(vl_zip, vh_zip, cl_zip, ch_zip)

Recombinant Sequence (light): DIQMTQSPSSLSASVGDRVTITCRASQDVNTAVAWYQQKPGKAPKLLIYSASFLYSGVPSRFSGSRSGTDFTLTISSLQPEDFATYYCQQHYTTPPTFGQGTKVEIKRTVAAPSVFIFPPSDEQLKSGTASVVCLLNNFYPREAKVQWKVDNALQSGNSQESVTEQDSKDSTYSLSSTLTLSKADYEKHKVYACEVTHQGLSSPVTKSFNRGEC
Recombinant Sequence (heavy): EVQLVESGGGLVQPGGSLRLSCAASGFNIKDTYIHWVRQAPGKGLEWVARIYPTNGYTRYADSVKGRFTISADTSKNTAYLQMNSLRAEDTAVYYCSRWGGDGFYAMDYWGQGTLVTVSSATLVTVSSASTKGPSVFPLAPCSRSTSESTAALGCLVKDYFPEPVTVSWNSGALTSGVHTFPAVLQSSGLYSLSSVVTVPSSSLGTKTYTCNVDHKPSNTKVDK


In [14]:
def make_df_heavies(df_filtered: pd.DataFrame, recombinant_seq_heavy) -> pd.DataFrame:
    """Creates a new dataframe for heavy chain metadata from an already-filtered dataframe."""

    df_heavies = pd.DataFrame({
    "pdb": df_filtered["pdb"],
    "template": df_filtered["template"],
    "Hchain": df_filtered["Hchain"],
    "H_isotype_clean": df_filtered["H_isotype_clean"],
    "HC_species": df_filtered["HC_species"],
    "H_coordinate_seq": df_filtered["H_coordinate_seq"],
    })

    # Clean Hchain to keep only first character
    df_heavies["Hchain"] = df_heavies["Hchain"].str[0]
    
    first_residue_nos = []
    last_residue_nos = []

    # Using iterrows for stable behaviour in pandas
    for idx, row in df_filtered.iterrows():

        # Split the string of numbering to a list and extract start/end values
        # Remove duplicate numbering with letter annotations eg: "100A"
        residue_numbering_list = [int(res) for res in row["H_PDB_numbering"].split(",") if res.isdigit()]
        last_residue_nos.append(residue_numbering_list[-1])
        first_residue_nos.append(residue_numbering_list[0])
    
    #Add new columns to new df
    df_heavies["H_chain_first_residue"] = first_residue_nos
    df_heavies["H_chain_last_residue"] = last_residue_nos

    # Append recombinant hybrid row to the new dataframe
    hybrid_row = {
        "pdb": "Fab_hybrid",
        "template": "target",
        
        # Assigns the constant region's isotype as the isotype for the target
        "H_isotype_clean": df_filtered.loc[df_filtered["template"] == "c_template", "H_isotype_clean"].values[0],

        # Assigns the species
        "HC_species": df_filtered.loc[df_filtered["template"] == "c_template", "HC_species"].values[0],
        
        # Add the recombinant sequence to the new row dict
        "H_coordinate_seq": recombinant_seq_heavy
    }
    
    df_heavies = pd.concat([df_heavies, pd.DataFrame([hybrid_row])], ignore_index=True)

    return df_heavies

df_heavy = make_df_heavies(df_filtered, recombinant_seq_heavy)
df_heavy

pdb    template Hchain H_isotype_clean    HC_species  \
0        1n8z  v_template      B            IgG1  homo_sapiens   
1        3ncj  c_template      H            IgG4  homo_sapiens   
2  Fab_hybrid      target    NaN            IgG4  homo_sapiens   

                                    H_coordinate_seq  H_chain_first_residue  \
0  EVQLVESGGGLVQPGGSLRLSCAASGFNIKDTYIHWVRQAPGKGLE...                    1.0   
1  EVQLVQSGAEVKKPGESLKISCKGSGYSFTNYWISWVRQMPGKGLE...                    1.0   
2  EVQLVESGGGLVQPGGSLRLSCAASGFNIKDTYIHWVRQAPGKGLE...                    NaN   

   H_chain_last_residue  
0                 220.0  
1                 217.0  
2                   NaN

In [15]:
def make_df_lights(df_filtered: pd.DataFrame, recombinant_seq_light) -> pd.DataFrame:
    """Creates a new dataframe for light chain metadata from an already-filtered dataframe."""

    df_lights = pd.DataFrame({
    "pdb": df_filtered["pdb"],
    "template": df_filtered["template"],
    "Lchain": df_filtered["Lchain"],
    "H_isotype_clean": df_filtered["H_isotype_clean"],
    "HC_species": df_filtered["HC_species"],
    "L_coordinate_seq": df_filtered["L_coordinate_seq"],
    })

    # Clean Hchain to keep only first character
    df_lights["Lchain"] = df_lights["Lchain"].str[0]
    
    first_residue_nos = []
    last_residue_nos = []

    # Using iterrows for stable behaviour in pandas
    for idx, row in df_filtered.iterrows():

        # Split the string of numbering to a list and extract start/end values
        # Remove duplicate numbering with letter annotations eg: "100A"
        residue_numbering_list = [int(res) for res in row["L_PDB_numbering"].split(",") if res.isdigit()]
        last_residue_nos.append(residue_numbering_list[-1])
        first_residue_nos.append(residue_numbering_list[0])
    
    #Add new columns to new df
    df_lights["L_chain_first_residue"] = first_residue_nos
    df_lights["L_chain_last_residue"] = last_residue_nos

    # Append recombinant hybrid row to the new dataframe
    hybrid_row = {
        "pdb": "Fab_hybrid",
        "template": "target",
        
        # Assigns the constant region's isotype as the isotype for the target
        "H_isotype_clean": df_filtered.loc[df_filtered["template"] == "c_template", "H_isotype_clean"].values[0],
        
        # Assigns the species
        "HC_species": df_filtered.loc[df_filtered["template"] == "c_template", "HC_species"].values[0],
        
        # Add the recombinant sequence to the new row dict
        "L_coordinate_seq": recombinant_seq_light
    }
    
    df_lights = pd.concat([df_lights, pd.DataFrame([hybrid_row])], ignore_index=True)
    
    return df_lights

df_light = make_df_lights(df_filtered, recombinant_seq_light)
df_light

pdb    template Lchain H_isotype_clean    HC_species  \
0        1n8z  v_template      A            IgG1  homo_sapiens   
1        3ncj  c_template      L            IgG4  homo_sapiens   
2  Fab_hybrid      target    NaN            IgG4  homo_sapiens   

                                    L_coordinate_seq  L_chain_first_residue  \
0  DIQMTQSPSSLSASVGDRVTITCRASQDVNTAVAWYQQKPGKAPKL...                    1.0   
1  DIQMTQSPSSLSASVGDRVTITCRASQSIGLYLAWYQQKPGKAPKL...                    1.0   
2  DIQMTQSPSSLSASVGDRVTITCRASQDVNTAVAWYQQKPGKAPKL...                    NaN   

   L_chain_last_residue  
0                 214.0  
1                 214.0  
2                   NaN

In [16]:
def write_fastas(df_l, df_h, fasta_out_dir):
    """Write 2 fasta files for light and heavy chains."""
    import os
    from datetime import datetime
    
    # Create date-time stamp to add to file name
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    
    # Create isotype label to add to the filename
    isotype = str(df_filtered.query("template == 'c_template'")["H_isotype_clean"].iloc[0])

    try:
        # ===Light FASTA===
        lfilename = f"light_chain_{isotype}_{timestamp}.fasta"
        stamped_l_file = os.path.join(fasta_out_dir, lfilename)
        with open(stamped_l_file, "w") as lf:
            for _, row in df_l.iterrows():

                # Skip fields that are NA
                if pd.isna(row["Lchain"]):
                    chain = ""
                else:
                    chain = f"Chain {row["Lchain"]}|"

                header = (
                    f'>{row["pdb"]}|{chain}Light_Chain_Fab_{row["H_isotype_clean"]} {row["template"]}|'
                    f'{row["HC_species"] if pd.notna(row["HC_species"]) else "species not listed"}'
                )
                sequence = row["L_coordinate_seq"]

                lf.write(header + "\n" + sequence + "\n")

        # ===Heavy FASTA===
        hfilename = f"heavy_chain_{isotype}_{timestamp}.fasta"
        stamped_h_file = os.path.join(fasta_out_dir, hfilename)
        with open(stamped_h_file, "w") as hf:
            for _, row in df_h.iterrows():

                # Skip fields that are NA
                if pd.isna(row["Hchain"]):
                    chain = ""
                else:
                    chain = f"Chain {row["Hchain"]}|"

                header = (
                    f'>{row["pdb"]}|{chain}Heavy_Chain_Fab_{row["H_isotype_clean"]} {row["template"]}|'
                    f'{row["HC_species"] if pd.notna(row["HC_species"]) else "species not listed"}'
                )
                sequence = row["H_coordinate_seq"]

                hf.write(header + "\n" + sequence + "\n")


    except Exception as e:
        print(f"Error whilst writing FASTA file: {e}")

import my_run_info
write_fastas(df_light, df_heavy, my_run_info.fasta_out_dir)

In [17]:
# Combine heavy and light chain dataframes into one for .pir creation.
merged_df = pd.merge(df_light, df_heavy, on=["pdb", "template", "H_isotype_clean", "HC_species"])
merged_df

# Filter redundant columns
df_combined = merged_df[[
    "pdb", "template", "Lchain", "Hchain", "L_chain_first_residue", 
    "H_chain_first_residue", "L_chain_last_residue", "H_chain_last_residue"
]].copy()

# Add empty columns
df_combined["light_chain_is_first"] = None
df_combined["start_point"] = None
df_combined["start_letter"] = None
df_combined["gapped_seq_1"] = None
df_combined["end_point"] = None
df_combined["end_letter"] = None
df_combined["gapped_seq_2"] = None

df_combined

pdb    template Lchain Hchain  L_chain_first_residue  \
0        1n8z  v_template      A      B                    1.0   
1        3ncj  c_template      L      H                    1.0   
2  Fab_hybrid      target    NaN    NaN                    NaN   

   H_chain_first_residue  L_chain_last_residue  H_chain_last_residue  \
0                    1.0                 214.0                 220.0   
1                    1.0                 214.0                 217.0   
2                    NaN                   NaN                   NaN   

  light_chain_is_first start_point start_letter gapped_seq_1 end_point  \
0                 None        None         None         None      None   
1                 None        None         None         None      None   
2                 None        None         None         None      None   

  end_letter gapped_seq_2  
0       None         None  
1       None         None  
2       None         None

In [18]:
# access the Ig label of the c_template row["pdb"] and add it to
# the file path
# NOTE: add this to the write FASTA function
for idx, row in df_filtered.iterrows():
    if row["pdb"] == c_template:
        isotype_label = str(row["H_isotype_clean"])
        
print(isotype_label)

IgG4


In [ ]:
# Dont run this cell. THIS DOES NOT WORK because SeqRecords class object can only hold .id, .desc and .seq.
from Bio import SeqIO

records = SeqIO.parse("THIS_IS_YOUR_INPUT_FILE.clustal", "clustal")
count = SeqIO.write(records, "THIS_IS_YOUR_OUTPUT_FILE.pir", "pir")
print("Converted %i records" % count)

In [19]:
from Bio.PDB.MMCIFParser import MMCIFParser
import os

def cif_parse(pdb: str, cif_fname: str, filepath: str):
    """Parses .cif and determines correct chain order for MODELLER."""
    from Bio.PDB.MMCIFParser import MMCIFParser
    import os

    # Set full path
    full_cif_path = os.path.join(filepath, cif_fname)

    # Initialise parser
    parser = MMCIFParser(QUIET=True)

    # Load structure
    structure = parser.get_structure(pdb, full_cif_path)

    # Extract first model (usually there's only one)
    model = next(structure.get_models())

    # Get chains in order
    cif_chain_order = [chain.id for chain in model]

    print(cif_chain_order)

    return cif_chain_order

import my_run_info
v_cif_chain_order = cif_parse(v_template, f"{v_template}.cif", my_run_info.cif_dir)
c_cif_chain_order = cif_parse(c_template, f"{c_template}.cif", my_run_info.cif_dir)

['A', 'B', 'C']
['L', 'H']


In [ ]:
# Redundant
def relevant_chains(pdb, df, cif_chain_order):
    """Extracts relevant chains from dataframes"""

    # Using iterrows for stable behaviour in pandas
    for idx, row in df.iterrows():
        if row["pdb"] == pdb:
            l_chain_letter = row["Lchain"]
            h_chain_letter = row["Hchain"]

            relevant_chains = [chain_id for chain_id in cif_chain_order if chain_id in (h_chain_letter, l_chain_letter)]
            print(relevant_chains)

            if relevant_chains[0] == l_chain_letter:
                start_chain = l_chain_letter
                start_point = int(row["L_chain_first_residue"])
                end_chain = h_chain_letter
                end_point = int(row["H_chain_last_residue"])

            if relevant_chains[0] == h_chain_letter:
                start_chain = h_chain_letter
                start_point = int(row["H_chain_last_residue"])
                end_chain = l_chain_letter
                end_point = int(row["L_chain_first_residue"])

    return (start_point, start_chain, end_point, end_chain)


relevant_chains(v_template, df_combined, v_cif_chain_order)

['A', 'B']


(1, 'A', 220, 'B')

In [26]:
def relevant_chains(df, v_cif_chain_order, c_cif_chain_order):
    """Extracts relevant chains from dataframes"""

    # Using iterrows for stable behaviour in pandas
    for idx, row in df.iterrows():
        if row["pdb"] == v_template:
            l_chain_letter = row["Lchain"]
            h_chain_letter = row["Hchain"]

            v_relevant_chains = [chain_id for chain_id in v_cif_chain_order if chain_id in (h_chain_letter, l_chain_letter)]
            print(v_relevant_chains)

            if v_relevant_chains[0] == l_chain_letter:
                # assign True/False to light_chain_is_first column
                df.at[idx, "light_chain_is_first"] = True
                df.at[idx, "start_point"] = int(row["L_chain_first_residue"])
                df.at[idx, "start_letter"] = row["Lchain"]
                df.at[idx, "end_point"] = int(row["H_chain_last_residue"])
                df.at[idx, "end_letter"] = row["Hchain"]
            else:
                df.at[idx, "light_chain_is_first"] = False
                df.at[idx, "start_point"] = int(row["H_chain_first_residue"])
                df.at[idx, "start_letter"] = row["Hchain"]
                df.at[idx, "end_point"] = int(row["L_chain_last_residue"])
                df.at[idx, "end_letter"] = row["Lchain"]

        if row["pdb"] == c_template:
            l_chain_letter = row["Lchain"]
            h_chain_letter = row["Hchain"]

            c_relevant_chains = [chain_id for chain_id in c_cif_chain_order if chain_id in (h_chain_letter, l_chain_letter)]
            print(c_relevant_chains)

            if c_relevant_chains[0] == l_chain_letter:
                df.at[idx, "light_chain_is_first"] = True
                df.at[idx, "start_point"] = int(row["L_chain_first_residue"])
                df.at[idx, "start_letter"] = row["Lchain"]
                df.at[idx, "end_point"] = int(row["H_chain_last_residue"])
                df.at[idx, "end_letter"] = row["Hchain"]
            else:
                df.at[idx, "light_chain_is_first"] = False
                df.at[idx, "start_point"] = int(row["H_chain_first_residue"])
                df.at[idx, "start_letter"] = row["Hchain"]
                df.at[idx, "end_point"] = int(row["L_chain_last_residue"])
                df.at[idx, "end_letter"] = row["Lchain"]
    return df

df_combined_populated = relevant_chains(df_combined, v_cif_chain_order, c_cif_chain_order)
df_combined_populated


['A', 'B']
['L', 'H']


pdb    template Lchain Hchain  L_chain_first_residue  \
0        1n8z  v_template      A      B                    1.0   
1        3ncj  c_template      L      H                    1.0   
2  Fab_hybrid      target    NaN    NaN                    NaN   

   H_chain_first_residue  L_chain_last_residue  H_chain_last_residue  \
0                    1.0                 214.0                 220.0   
1                    1.0                 214.0                 217.0   
2                    NaN                   NaN                   NaN   

  light_chain_is_first start_point start_letter  \
0                 True           1            A   
1                 True           1            L   
2                 None        None         None   

                                        gapped_seq_1 end_point end_letter  \
0  DIQMTQSPSSLSASVGDRVTITCRASQDVNTAVAWYQQKPGKAPKL...       220          B   
1  DIQMTQSPSSLSASVGDRVTITCRASQSIGLYLAWYQQKPGKAPKL...       217          H   
2  DIQMTQSPSSLSASVGDRVTITCRASQDVNTAVAWYQQKPGKAPKL...      None       None   

                                        gapped_seq_2  
0  EVQLVESGGGLVQPGGSLRLSCAASGFNIKDTYIHWVRQAPGKGLE...  
1  EVQLVQSGAEVKKPGESLKISCKGSGYSFTNYWISWVRQMPGKGLE...  
2  EVQLVESGGGLVQPGGSLRLSCAASGFNIKDTYIHWVRQAPGKGLE...

In [27]:
import my_run_info

def extract_gapped_seqs(df, msa_filepath, light_clustal_fname, heavy_clustal_fname):
    """Parses .aln-clustal files with Bio.SeqIO and extracts the gapped sequences."""

    from Bio import SeqIO
    from Bio import AlignIO
    import os

    # Read clustal files
    def load_gapped_seqs(filename):
        full_clustal_path = os.path.join(msa_filepath, filename)
        alignment = AlignIO.read(full_clustal_path, "clustal")

        # Dictionary of alignments using a for loop to return {ID: seq...}
        return {record.id.split("|")[0].lower(): str(record.seq) for record in alignment}

    # Load both alignments
    light_seqs = load_gapped_seqs(light_clustal_fname)
    heavy_seqs = load_gapped_seqs(heavy_clustal_fname)
    
    # Assign sequences to gapped_seq_1 and gapped_seq_2 if 
    # light_chain_is_first is True, otherwise 2 then 1
    for idx, row in df.iterrows():
        pdb = row["pdb"].lower()
        if row["light_chain_is_first"] is True:
            df.at[idx, "gapped_seq_1"] = light_seqs.get(pdb, "").replace("\n", "")
            df.at[idx, "gapped_seq_2"] = heavy_seqs.get(pdb, "").replace("\n", "")
        elif row["light_chain_is_first"] is False:
            df.at[idx, "gapped_seq_1"] = heavy_seqs.get(pdb, "").replace("\n", "")
            df.at[idx, "gapped_seq_2"] = light_seqs.get(pdb, "").replace("\n", "")

    # Handle Fab_hybrid row manually by finding the target row
    hybrid_idx = df[df["pdb"].str.lower() == "fab_hybrid"].index
    if not hybrid_idx.empty:
        idx = hybrid_idx[0]
        df.at[idx, "gapped_seq_1"] = light_seqs.get("fab_hybrid", "").replace("\n", "")
        df.at[idx, "gapped_seq_2"] = heavy_seqs.get("fab_hybrid", "").replace("\n", "")


    return df

df_for_pir = extract_gapped_seqs(
    df_combined_populated,
    my_run_info.clustal_out_dir,
    light_clustal_fname="Fab_alignment_IgG4_light.aln-clustal", # replace this with automation
    heavy_clustal_fname="Fab_alignment_IgG4_heavy.aln-clustal"
)
df_for_pir

pdb    template Lchain Hchain  L_chain_first_residue  \
0        1n8z  v_template      A      B                    1.0   
1        3ncj  c_template      L      H                    1.0   
2  Fab_hybrid      target    NaN    NaN                    NaN   

   H_chain_first_residue  L_chain_last_residue  H_chain_last_residue  \
0                    1.0                 214.0                 220.0   
1                    1.0                 214.0                 217.0   
2                    NaN                   NaN                   NaN   

  light_chain_is_first start_point start_letter  \
0                 True           1            A   
1                 True           1            L   
2                 None        None         None   

                                        gapped_seq_1 end_point end_letter  \
0  DIQMTQSPSSLSASVGDRVTITCRASQDVNTAVAWYQQKPGKAPKL...       220          B   
1  DIQMTQSPSSLSASVGDRVTITCRASQSIGLYLAWYQQKPGKAPKL...       217          H   
2  DIQMTQSPSSLSASVGDRVTITCRASQDVNTAVAWYQQKPGKAPKL...      None       None   

                                        gapped_seq_2  
0  EVQLVESGGGLVQPGGSLRLSCAASGFNIKDTYIHWVRQAPGKGLE...  
1  EVQLVQSGAEVKKPGESLKISCKGSGYSFTNYWISWVRQMPGKGLE...  
2  EVQLVESGGGLVQPGGSLRLSCAASGFNIKDTYIHWVRQAPGKGLE...

>P1;1n8z
structureX:1n8z:1:A:220:B:V_Var_Template:::
DIQMTQSPSSLSASVGDRVTITCRASQDV-NTAVAWYQQKPGKAPKLLIYSASFLYSGVP
SRFSGSRSGTDFTLTISSLQPEDFATYYCQQHYTTPPTFGQGTKVEIKRTVAAPSVFIFP
PSDEQLKSGTASVVCLLNNFYPREAKVQWKVDNALQSGNSQESVTEQDSKDSTYSLSSTL
TLSKADYEKHKVYACEVTHQGLSSPVTKSFNRGEC/EVQLVESGGGLVQPGGSLRLSCAA
SGFNIKD--TYIHWVRQAPGKGLEWVARIYPTNGYTRYADSVKGRFTISADTSKNTAYLQ
MNSLRAEDTAVYYCSRWGGDGFYAMDYWGQGTLVTVSSASTKGPSVFPLAPSSKSTSGGT
AALGCLVKDYFPEPVTVSWNSGAL--TSGVHTFPAVLQSSGLYSLSSVVTVPSSSLG---
TQTYICNVNHKPSNTKVDKKVEP--*


In [32]:
# The allowed format:
# >P1;3m8o
# structure:pdb_file:.:.:.:.::::
# seq1---/seq2---*

def write_modeller_pir(df: pd.DataFrame, out_path: str):
    """Writes a .pir file from a df containing gapped sequences"""
    
    with open(out_path, "w") as pir:
        for _, row in df.iterrows():
            if row["pdb"] == v_template:
                
                v_pir_header = (
                    f">P1;{row['pdb']}\n"
                    f"structureX:{row['pdb']}:{row["start_point"]}:{row["start_letter"]}:{row["end_point"]}:{row["end_letter"]}"
                    f":variable_template:::\n"
                )
                v_gapped_sequence = (f"{row['gapped_seq_1']}/{row['gapped_seq_2']}*\n")

            if row["pdb"] == c_template:

                c_pir_header = (
                    f">P1;{row['pdb']}\n"
                    f"structureX:{row['pdb']}:{row["start_point"]}:{row["start_letter"]}:{row["end_point"]}:{row["end_letter"]}"
                    f":constant_template_{isotype_label}:::\n"
                )
                c_gapped_sequence = (f"{row['gapped_seq_1']}/{row['gapped_seq_2']}*\n")

            if row["template"] == "target":

                target_pir_header = (
                    f">P1;{row['pdb']}_{isotype_label}_target\n"
                    f"sequence:{row['pdb']}_{isotype_label}_target::.::.:hybrid_{isotype_label}_target:::\n"
                )
                target_gapped_sequence = (f"{row['gapped_seq_1']}/{row['gapped_seq_2']}*")

        pir.write(v_pir_header + v_gapped_sequence)
        pir.write(c_pir_header + c_gapped_sequence)
        pir.write(target_pir_header + target_gapped_sequence)

write_modeller_pir(df_for_pir, "../pir_files/pir_alignment_IgG4.pir")